scylla/src/transport/locator/mod.rs

```diff
@@ -0,0 +1,522 @@
+            .precomputed_replicas
+            .get_precomputed_simple_strategy_replicas(token, replication_factor)
+        {
+            precomputed_replicas.into()
+        } else {
+            Replicas::from_iter(
+                self.replication_data
+                    .simple_strategy_replicas(token, replication_factor),
+            )
+        }
+    }
+
+    fn get_network_strategy_replicas<'a>(
+        &'a self,
+        token: Token,
+        datacenter: &str,
+        datacenter_replication_factor: usize,
+    ) -> Replicas<'a> {
+        if datacenter_replication_factor == 0 {
+            return EMPTY_REPLICAS;
+        }
+
+        if let Some(precomputed_replicas) = self
+            .precomputed_replicas
+            .get_precomputed_network_strategy_replicas(
+                token,
+                datacenter,
+                datacenter_replication_factor,
+            )
+        {
+            Replicas::from(precomputed_replicas)
+        } else {
+            Replicas::from_iter(self.replication_data.nts_replicas_in_datacenter(
+                token,
+                datacenter,
+                datacenter_replication_factor,
+            ))
+        }
+    }
+}
+
+#[derive(Debug)]
+enum ReplicaSetInner<'a> {
+    Plain(Replicas<'a>),
+
+    // Represents a set of SimpleStrategy replicas that is limited to a specified datacenter.
+    FilteredSimple {
+        replicas: Replicas<'a>,
+        datacenter: &'a str,
+    },
+
+    // Represents a set of NetworkTopologyStrategy replicas that is not limited to any specific
+    // datacenter. The set is constructed lazily, by invoking
+    // `locator.get_network_strategy_replicas()`.
+    MergedNTS {
+        datacenter_repfactors: &'a HashMap<String, usize>,
+        locator: &'a ReplicaLocator,
+        token: Token,
+    },
+}
+
+/// Represents a set of replicas for a given token and strategy;
+///
+/// This container can only be created by calling `ReplicaLocator::replicas_for_token`, and it
+/// can borrow precomputed replica lists living in the locator.
+#[derive(Debug)]
+pub struct ReplicaSet<'a> {
+    inner: ReplicaSetInner<'a>,
+}
+
+impl<'a> ReplicaSet<'a> {
+    /// Gets the nth replica.
+    /// Defined order does not have to match the one set by the token ring.
+    ///
+    /// If the set represents `SimpleStrategy` replicas that were filtered by datacenter, this
+    /// function will have O(R) complexity, where R is the replication factor of that strategy.
+    ///
+    /// If the set represents `NetworkTopologyStrategy` replicas that were not filtered by
+    /// datacenter, this function will have O(D) complexity where D is the number of known
+    /// datacenters. If additionally this strategy was not requested for precomputation upon
+    /// locator, this function will have O(D * N) complexity, where N - number of all nodes.
+    ///
+    /// In all other cases, the complexity is O(1)
+    pub fn get(&self, index: usize) -> Option<NodeRef<'a>> {
```

Comment:
I'm wondering if it makes sense for this method to exist. It has a weird complexity when `NetworkTopologyStrategy` replicas are not precomputed. It also seems strange to me to use indexing in a type which name ends with `Set`.
I can't think of a valid use case for invoking `get(3)` (except for choosing randomly).

Maybe a better idea would be to throw this method away and add a `pick_random(predicate)` that would be arguably more useful in load balancing policies creation context.

Reply:
Is this comment resolved?